src/bellperson/shape_cs.rs

```diff
@@ -0,0 +1,328 @@
+//! Support for generating R1CS shape using bellperson.
+
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, HashMap};
+
+use crate::traits::{Group, PrimeField as PF};
+use ff::{Field, PrimeField};
+
+use bellperson::{ConstraintSystem, Index, LinearCombination, SynthesisError, Variable};
+
+#[derive(Clone, Copy)]
+struct OrderedVariable(Variable);
+
+#[derive(Debug)]
+enum NamedObject {
+  Constraint(usize),
+  Var(Variable),
+  Namespace,
+}
+
+impl Eq for OrderedVariable {}
+impl PartialEq for OrderedVariable {
+  fn eq(&self, other: &OrderedVariable) -> bool {
+    match (self.0.get_unchecked(), other.0.get_unchecked()) {
+      (Index::Input(ref a), Index::Input(ref b)) => a == b,
+      (Index::Aux(ref a), Index::Aux(ref b)) => a == b,
+      _ => false,
+    }
+  }
+}
+impl PartialOrd for OrderedVariable {
+  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+    Some(self.cmp(other))
+  }
+}
+impl Ord for OrderedVariable {
+  fn cmp(&self, other: &Self) -> Ordering {
+    match (self.0.get_unchecked(), other.0.get_unchecked()) {
+      (Index::Input(ref a), Index::Input(ref b)) => a.cmp(b),
+      (Index::Aux(ref a), Index::Aux(ref b)) => a.cmp(b),
+      (Index::Input(_), Index::Aux(_)) => Ordering::Less,
+      (Index::Aux(_), Index::Input(_)) => Ordering::Greater,
+    }
+  }
+}
+
+#[allow(clippy::upper_case_acronyms)]
+/// TODO: document
```

Comment:
It would be great to remove all the TODOs

Reply:
Will do. I thought I had, but apparently not in this file.

Reply:
Done.